feat: add limitk() and limit_ratio() aggregation operators

jjo opened the pull request:
Implement `limitk(N, metric)` and `limit_ratio(R, metric)`:

* `limitk(N, metric)` returns the 1st N samples gathered from metric, _without_ any particular order
* `limit_ratio(R, metric)` with `0 <= R <= 1.0` returns approximately `R` ratio of samples

Sometimes you want to "sample" some timeseries, for example to explore
their labels names and values, currently the only operator available is
`topk(K, metric)` (and `bottomk()` of course), needing to "pay" for its
ordering feature by visiting all samples, and sorting the topk-s.

This change introduces `limitk(N, metric)` which _just_ samples
the first points visited without any further/guaranteed ordering, and
also `limit_ratio(R, metric)` for cases where you want a more "fair"
distribution of the samples, using `R` as a "proxy" for an approximate
sampling ratio, which uses `Metrics.Hash()/MaxUint64` as the sample
ratio, and select the sample if its ratio is lesser than `R`. 

An `promql/engine.go` excerpt to expand on `limit_ratio()`:
(NOTE: `sampleOffset()` returns `sample.Metric.Hash()) / MaxUint64`)
```
    // If ratioLimit >= 0: add sample if sampleOffset is lesser than ratioLimit
    //
    // 0.0        ratioLimit                1.0
    //  [---------|--------------------------]
    //  [#########...........................]
    //
    // e.g.:
    //   sampleOffset==0.3 && ratioLimit==0.4
    //     0.3 < 0.4 ? --> add sample
    //
    // Else if ratioLimit < 0: add sample if rand() return the "complement" of ratioLimit>=0 case
    // (loosely similar behavior to negative array index in other programming languages)
    //
    // 0.0       1+ratioLimit               1.0
    //  [---------|--------------------------]
    //  [.........###########################]
    //
    // e.g.:
    //   sampleOffset==0.3 && ratioLimit==-0.6
    //     0.3 >= 0.4 ? --> don't add sample
    sampleOffset := s.sampleOffset(sample.T, sample)
    return (ratioLimit >= 0 && sampleOffset < ratioLimit) ||
        (ratioLimit < 0 && sampleOffset >= (1.0+ratioLimit))
```

Testing
-------
* added `promql/promqltest/testdata/limit.test`
* added tests to `promql/bench_test.go`

Web
---
* added `web/ui/` artifacts

NB: `sample_ratio()` was previously `sample_random()`, changed to a deterministic implementation, as per PR discussion.
NB: `sample_limit()` and `sample_ratio()` were renamed to `limitk()` and `limit_ratio()` respectively


Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

roidelapluie commented:
Thanks for your pull request. Deterministic results are a key factor of PromQL, so limit_random does not fit. We decided in our last dev summit to implement limit, but at the API level, for all API endpoints. Not in PromQL.

juliusv commented:
> Thanks for your pull request. Deterministic results are a key factor of PromQL, so limit_random does not fit. We decided in our last dev summit to implement limit, but at the API level, for all API endpoints. Not in PromQL.

To add to that, we decided to make the API-level limits non-deterministic to begin with (i.e. it's ok to return an arbitrary subset), but might potentially make them more deterministic in the future.

But your PR's motivations does raise the questions:

* If such a limit is applied overall to a PromQL query, can we make it affect the PromQL evaluation in such a way that we don't have to compute all the underlying results if most of them are thrown away anyway (important only for queries selecting huge numbers of metrics). At least for bare selectors without anything wrapping them, this should be an easy thing to do.
* Should we add an option to make the results *explicitly* random. I guess your hope for that would be to possibly get a more "fair" look at the data e.g. if the ordered results are always not representative in the way you care about, but randomness of course can also never really guarantee anything, so I'm not sure about that yet.

jjo commented:
> Thanks for your pull request. Deterministic results are a key factor of PromQL, so limit_random does not fit.

Thanks for the reply, I ab-used the `Randomizer` interface to implement a non-random/deterministic
HashRandomizer that returns `[0.0, 1.0]` as `labels.Hash() / MaxUint64` at 181b34b

> We decided in our last dev summit to implement limit, but at the API level, for all API endpoints. Not in PromQL.

IMO there may still be use cases for limit at PromQL, it would also not need any UI artifact to be used (like e.g. an input field to specific such limit value).

jjo commented:
> > Thanks for your pull request. Deterministic results are a key factor of PromQL, so limit_random does not fit. We decided in our last dev summit to implement limit, but at the API level, for all API endpoints. Not in PromQL.
> 
> To add to that, we decided to make the API-level limits non-deterministic to begin with (i.e. it's ok to return an arbitrary subset), but might potentially make them more deterministic in the future.
> 
> But your PR's motivations does raise the questions:
> 
> * If such a limit is applied overall to a PromQL query, can we make it affect the PromQL evaluation in such a way that we don't have to compute all the underlying results if most of them are thrown away anyway (important only for queries selecting huge numbers of metrics). At least for bare selectors without anything wrapping them, this should be an easy thing to do.

If I get you correctly, I've already implemented it at (i.e. `continue` after filling the heap with `k` elements): 
https://github.com/jjo/prometheus/blob/181b34b9a921cbf38040feb2114d3a1f7cdfb338/promql/engine.go#L2763-L2772

> * Should we add an option to make the results _explicitly_ random. I guess your hope for that would be to possibly get a more "fair" look at the data e.g. if the ordered results are always not representative in the way you care about, but randomness of course can also never really guarantee anything, so I'm not sure about that yet.

Actually I went the order way, adding a deterministic HashRandomizer (please mind the Rand in the name :P), that returns a float64 in `[0.0, 0.1]` range as `labels.Hash() / MaxUint64`, the used it as implementation of the `Randomizer` interface I had added, see 181b34b.

As an interesting feature, also added the possibility to specify _negative_ `p` numbers in `sample_random(-p, metric)`, which will return "the complement" (set of timeseries) than the `sample_random(p, metric)` counterpart, excerpt from the code:

```
    case q >= 0:            
        // If q >= 0 add sample if randomizer.Float64() is lesser than q
        //                                                                    
        // 0.0        q                         1.0                                      
        //  [---------|--------------------------]    
        //  [#########...........................]
        //   
        // e.g.:                                                          
        //   randomizer.Float64()==0.3 && q==0.4                          
        //     0.3 < 0.4 ? --> add sample                    
        //                                                                                                      
        if randomizer.Float64(ts, sample) < q {                                        
            heap.Push(h, sample)            
        }                                 
    case q < 0:                                   
        // If q < 0 add sample if rand() return the "complement" of q>=0 case
        //                                  
        // 0.0       1+q                        1.0
        //  [---------|--------------------------] 
        //  [.........###########################]          
        //                                      
        // e.g.:                                      
        //   randomizer.Float64()==0.3 && q==-0.6        
        //     0.7 >= 0.6 ? --> don't add sample                  
        if randomizer.Float64(ts, sample) >= (1.0 + q) {
            heap.Push(h, sample)                           
        }                                           
    }                                          
```

This feature could also open some interesting use cases, as it allows subsetting the timeseries by a "pivot" value, returning deterministic "complement"  of timeseries (depending on their labels' hash).

jjo commented:
> Actually, I went the order way, adding a deterministic HashRandomizer (please mind the Rand in the name :P), that returns a float64 in `[0.0, 0.1]` range as `labels.Hash() / MaxUint64`, the used it as implementation of the `Randomizer` interface I had added, see [181b34b](https://github.com/prometheus/prometheus/commit/181b34b9a921cbf38040feb2114d3a1f7cdfb338).
> 
> As an interesting feature, also added the possibility to specify _negative_ `p` numbers in `sample_random(-p, metric)`, which will return "the complement" (set of timeseries) than its `sample_random(p, metric)` counterpart, excerpt from the code:

@juliusv you can see this "in action" at the [promql/engine_test.go test case](https://github.com/jjo/prometheus/blob/181b34b9a921cbf38040feb2114d3a1f7cdfb338/promql/engine_test.go#L1724-L1767) I added.

roidelapluie commented:
We have discussed this during our weekly bug scrub.

We would like to better understand the gap between implementing this in PromQL and the implementation of a limit directly at the API level, maybe in a broader design document gathering use cases.

For now, as we agreed with implementing this at the API level, we will leave this PR open until we can better perceive the added value of this pull request.

roidelapluie commented:
We will also discuss this topic at our next Dev Summit in Berlin, at the end of september.

jjo commented:
> We will also discuss this topic at our next Dev Summit in Berlin, at the end of September.
> 
> Thanks!

Thanks Julien!, appreciate the details and context. :) 

Note that API limit would be the ~equivalent of `sample_limit()`, but `sample_ratio(p, metrics)` (as per my update above, using `labels Hash()/MaxUint64` to compare against against `p`) is quite different.

Do you think it would be worth fixing this PR to narrow it to `sample_ratio()` implementation only?
Can also work on a design/use-case docu if needed.

beorn7 commented:
Sadly, we still haven't discussed this at the dev summit, see [the relevant agenda item](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?pli=1#bookmark=id.pr0js4n43f6z).

On a more general not, we now have [a feature flag for experimental PromQL functions](https://github.com/prometheus/prometheus/pull/13103), so maybe just put this behind the flag, and we can merge it without a super deep discussion? @roidelapluie WDYT?

roidelapluie commented:
Yes, let's do it behind the flag.

jjo commented:
> Yes, let's do it behind the flag.

Thanks folks for the feedback, will do.
/cc @beorn7

bwplotka commented:
We have consensus to have this behind flag (on DevSummit), but I think we have some suggestions for the name. To me it seems this mimics "topk" just without priority queue. Topk is kind of "series" limit not sample 🙈 thus "sample_limit" might be ambiguous -- does it mean vertical samples (series) or horizontal (e.g. with over_time).

cc @bboreham @jan--f had some opinions.

beorn7 commented:
@roidelapluie @juliusv @bwplotka any chance any of you could do the code level review here?

zeitlinger commented:
I have a use case involving `sort_by_label`:

`sample_limit(1, sort_by_label(m, l))`

I want to sort the metric expression `m` by label `l` and then get the sample with the smallest value of l.

This addresses a use case where I have multiple metrics that potentially fit what I'm looking for.
Similar to using `OR`, but the query side doesn't need to be aware of the metric names.
What I want to do instead is add label l at ingest time (using a relabel rule).

beorn7 reviewed (commented):
A discussion with @zeitlinger and @bboreham about a specific use case (see @zeitlinger's comment) brought this back to my attention.

I'm very sorry that this has been in review limbo for so long. I'm lacking time and competence to properly review this (even though I just found a little detail, see comment below). The more qualified people I have tried to rally haven't even responded to my ping above (not even with "sorry, I won't have time for this any time soon"). Let's discuss our new ideas, and then let's get this reviewed and merged ASaP.

beorn7 commented:
So here some more thoughts about @zeitlinger's use case:

The `sort_by_label` is pretty new, so arguably this use case didn't even exist when this PR was started.

As we had only `sort` and `sort_desc` initially, it made sense to have the sister functions `topk` and `bottomk`. Now we have `sort_by_label` and `sort_by_label_desc` but instead of creating `topk_by_label` and `bottomk_by_label`, it would be more elegant to simply piggyback on the `sample_limit` proposed here. But of course, that means `sample_limit` has to preserve the sorting within the instant vector. I hope that's feasible to implement without huge efficiency loss.

About the naming: A agree with @bwplotka that "sample" is problematic as a term. It just has too many meanings in the context of PromQL. How about renaming `sample_limit` to `limitk`? That corresponds nicely to `topk` and `bottomk`, and as a bonus, it's the same name as in [Victoria Metrics](https://docs.victoriametrics.com/metricsql/#limitk). Maybe the function even works the same way. I'm not sure if "The returned set of time series remain the same across calls." implies that it keeps the order of the input vector.

Which leaves us with a maybe better name for `sample_ratio`. Although that name makes sense in another way: The function picks a (pseudo random) sample of vector elements according to the provided ratio.

bboreham commented:
Hello from the bug-scrub!

The aggregation functions have been rewritten since this PR was opened, which means there is quite a lot of work to re-integrate this change.

Thinking about the implementation, in view of @zeitlinger's comment (and @bwplotka said this too), `limitk` needs to operate on the series and not on the samples. I.e. it needs to pick one set of series in the order output from the previous function, and use those consistently across the range-query.  I don't think any of the existing aggregations works like this.  Structurally it could be like `label_replace` and `label_join`, which operate on series not on samples.

@jjo can you let us know whether you have time to do such a large rework, or should we look for someone else to try?

beorn7 commented:
> it needs to pick one set of series in the order output from the previous function, and use those consistently across the range-query

:thinking: This sounds like it would break the semantics of a range query (i.e. a range query must give the same result as a series of instant queries with the corresponding evaluation timestamps – which is something we (deliberately) broke by (cautiously) introducing `start()` and `end()` as something you could use after `@`).

So maybe it's enough to require that `limitk` must act on the order of its input.

Example: 

`limitk by (instance) (1, sort_by_label(temperature_kelvin, "priority"))` returns the lowest priority temperature measurement from each instance. Series with a given priority might come and go, but at each step in the range query, I want to see the one with the lowest priority value, which could be a different one at different times (rather than seeing nothing at some evaluation times if the lowest-priority series over the whole range isn't present at all evaluation times).

If you want to see the same series over a range, you can do the same trick as with `topk`, i.e. `temperature_kelvin and limitk by (instance) (1, sort_by_label(temperature_kelvin @ end(), "priority"))`.

It would help, though, if `sort_by_label` consistently returned the same order of metrics. Right now it seems to fall back to sorting by sample value if the label value to sort by is the same. Maybe it should fall back to the metric hash so that the output order is always the same, even if the sample value changes.

jjo commented on promql/engine.go:
```diff
@@ -2649,6 +2649,17 @@ func (ev *evaluator) aggregation(e *parser.AggregateExpr, grouping []string, par
+		switch {
+		case q == 0:
+			return Vector{}, annos
+		case q > 1.0:
+			ev.errorf("Float value %v is greater than 1.0", q)
+		case q < -1.0:
+			ev.errorf("Float value %v is less than -1.0", q)
```
+1 implementing in next commit (after full rebasing, many `main` changes since)

jjo replied:
pushed on latest  change(s)

jjo commented:
> So here some more thoughts about @zeitlinger's use case:
> 
> The `sort_by_label` is pretty new, so arguably this use case didn't even exist when this PR was started.
> 
> As we had only `sort` and `sort_desc` initially, it made sense to have the sister functions `topk` and `bottomk`. Now we have `sort_by_label` and `sort_by_label_desc` but instead of creating `topk_by_label` and `bottomk_by_label`, it would be more elegant to simply piggyback on the `sample_limit` proposed here. But of course, that means `sample_limit` has to preserve the sorting within the instant vector. I hope that's feasible to implement without huge efficiency loss.
> 
> About the naming: A agree with @bwplotka that "sample" is problematic as a term. It just has too many meanings in the context of PromQL. How about renaming `sample_limit` to `limitk`? That corresponds nicely to `topk` and `bottomk`, and as a bonus, it's the same name as in [Victoria Metrics](https://docs.victoriametrics.com/metricsql/#limitk). Maybe the function even works the same way. I'm not sure if "The returned set of time series remain the same across calls." (quote from ViMe documentation) implies that it keeps the order of the input vector.
> 
> Which leaves us with a maybe better name for `sample_ratio`. Although that name makes sense in another way: The function picks a (pseudo random) sample of vector elements according to the provided ratio.

Agreed on the renaming!,  `limitk` is much more natural, while leaving a "nice ambiguity" regarding being sample- xor series- wise 😇 , FWIW I essentially followed the same [non/]assumptions for returned series as `topk` and `bottomk` siblings

jjo committed:
rebase 2024-07-01, picks previous renaming to `limitk()` and `limit_ratio()`

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> Yes, let's do it behind the flag.

Next in line, after reworking the whole PR for recent `main` changes, and renaming them to `limitk()` and `limit_ratio()`

jjo committed:
gofumpt -d -extra

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
more lint fixes

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
more lint fixes+

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
put limitk() and limit_ratio() behind --enable-feature=promql-experimental-functions

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> Yes, let's do it behind the flag.

Implemented at 89cea51, behind `--enable-feature=promql-experimental-functions`

jjo commented:
> Hello from the bug-scrub!
> 
> The aggregation functions have been rewritten since this PR was opened, which means there is quite a lot of work to re-integrate this change.
> 
> Thinking about the implementation, in view of @zeitlinger's comment (and @bwplotka said this too), `limitk` needs to operate on the series and not on the samples. I.e. it needs to pick one set of series in the order output from the previous function, and use those consistently across the range-query. I don't think any of the existing aggregations works like this. Structurally it could be like `label_replace` and `label_join`, which operate on series not on samples.
> 
> @jjo can you let us know whether you have time to do such a large rework, or should we look for someone else to try?

Thanks Bryan for the feedback!, I don't think I'd have time _and_ codebase knowledge to implement such core change for samples -> series approach, FWIW I "rebased" (actually fully reworked) previous changes against `main`, included renaming them to `limitk()` and `limit_ratio()`.

jjo committed:
EnableExperimentalFunctions for TestConcurrentRangeQueries() also

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> > Hello from the bug-scrub!
> > The aggregation functions have been rewritten since this PR was opened, which means there is quite a lot of work to re-integrate this change.
> > Thinking about the implementation, in view of @zeitlinger's comment (and @bwplotka said this too), `limitk` needs to operate on the series and not on the samples. I.e. it needs to pick one set of series in the order output from the previous function, and use those consistently across the range-query. I don't think any of the existing aggregations works like this. Structurally it could be like `label_replace` and `label_join`, which operate on series not on samples.
> > @jjo can you let us know whether you have time to do such a large rework, or should we look for someone else to try?
> 
> Thanks Bryan for the feedback!, I don't think I'd have time _and_ codebase knowledge to implement such core change for samples -> series approach, FWIW I "rebased" (actually fully reworked) previous changes against `main`, included renaming them to `limitk()` and `limit_ratio()`.

@bboreham thinking about switching to operate on series, I think it'd be tricky to choose between "strict" and somehow "best-effort" series-mode -> 
Imagine a metric like `kube_pod_info()` (which can potentially have a high TS churn rate), if we queried for e.g. `limitk(10, kube_pod_info{...}` but for the range_query time we had zero "sustained" TS (due to Pod recycling)
-> should we:
* 1) return empty? ie strict TS: there's none during the whole interval
* 2) return partial? ie strict TS: only some during initial period, zero afterwards
* 3) return all from 2) + sample- based? ie best-effort: 2) + sample- based

IMO I find it hard to choose between the above (and possible other alternatives), even trying to derive from the principle of least surprise, while also factoring existing `topk()` and `bottomk()` as samples- based.
Disclaimer: all the above assuming I properly understood `engine.go` in this my 1st codebase involvement 😇 .

jjo committed:
use testutil.RequireEqual to fix tests, WIP equivalent thingie for require.Contains

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
lint fix

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
moar linting

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
rebase 2024-06-19

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
re-add limit(2, metric) testing for N=2 common series subset

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
move `ratio = param` to default switch case, for better readability

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
gofumpt -d -extra util/testutil/cmp.go

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
early break when reaching k elems in limitk(), should have always been so (!)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
small typo fix

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
no-change small break-loop rearrange for readability

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
remove IsNan(ratio) condition in switch-case, already handled as input validation

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
no-change adding some comments

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
no-change simplify fullMatrix() helper functions used for tests

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
add `limitk(-1, metric)` testcase, which is handled as any k < 1 case

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
engine_test.go: no-change create `requireCommonSeries() helper func (moving code into it) for readability

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
rebase 2024-06-21

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
engine_test.go: HAPPY NOW about its code -> reorg, create and use simpleRangeQuery() function, less lines and more readable ftW \o/

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

beorn7 commented on promql/engine_test.go:
```diff
@@ -1745,6 +1749,262 @@ load 1ms
+		expectedLen, expected, actual))
+}
+
+func TestLimitK(t *testing.T) {
```
Could this (and TestLimitRatio below) be done via the PromQL testing framework? See all the files in https://github.com/prometheus/prometheus/tree/main/promql/promqltest/testdata . Could be a new file `limit.test` there.

beorn7 commented:
Thank you very much for reviving this.

I had a smartass comment about testing above (without having looked at the code in detail).

@bboreham agreed to have a closer look at the implementation.

jjo committed:
move limitk(), limit_ratio() testing to promql/promqltest/testdata/limit.test

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> Thank you very much for reviving this.
> 
> I had a smartass comment about testing above (without having looked at the code in detail).
> 
> @bboreham agreed to have a closer look at the implementation.

Thanks folks, moved testing to `testdata/` (great stuff btw!), and rebased to latest main.

jjo committed:
remove stale leftover after moving tests from engine_test.go to testdata/

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented on promql/engine.go:
```diff
@@ -2958,19 +2977,22 @@ func (ev *evaluator) aggregation(e *parser.AggregateExpr, q float64, inputMatrix
 }
 
-// aggregationK evaluates topk or bottomk at one timestep on inputMatrix.
+// aggregationK evaluates topk, bottomk, or limitk at one timestep on inputMatrix.
```
```suggestion
// aggregationK evaluates topk, bottomk, limitk, or limit_ratio at one timestep on inputMatrix.
```

juliusv commented on promql/engine.go:
```diff
@@ -2958,19 +2977,22 @@ func (ev *evaluator) aggregation(e *parser.AggregateExpr, q float64, inputMatrix
 // Output that has the same labels as the input, but just k of them per group.
 // seriesToResult maps inputMatrix indexes to groups indexes.
-// For an instant query, returns a Matrix in descending order for topk or ascending for bottomk.
+// For an instant query, returns a Matrix in descending order for topk or ascending for bottomk, or without any order for limitk.
```
```suggestion
// For an instant query, returns a Matrix in descending order for topk or ascending for bottomk, or without any order for limitk / limit_ratio.
```

jjo committed:
fix flaky `limit_ratio(0.5, ...)` test case

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented on promql/engine.go:
```diff
@@ -3419,6 +3478,65 @@ func makeInt64Pointer(val int64) *int64 {
+		//   ratiosampler.sampleOffset()==0.3 && ratioLimit==0.4
+		//     0.3 < 0.4 ? --> add sample
+		//
+		if sampleOffset < ratioLimit {
```
I realize this wasn't a `return` in past iterations, but with the current code structure it would probably be more readable to just return the result of the boolean comparison directly in both cases and not require the `return false` at the end at all.

jjo replied:
Oh good catch, fixed at 1188b10c.

juliusv commented on promql/engine.go:
```diff
@@ -3419,6 +3478,65 @@ func makeInt64Pointer(val int64) *int64 {
+
+func (s *HashRatioSampler) sampleOffset(ts int64, sample *Sample) float64 {
+	const (
+		float64MaxUint64 = float64(math.MaxUint64)
```
I guess this is ok in practice, although it makes me at least slightly uncomfortable that a float64 won't be able to represent anything above 2^53 without loss of precision (so you won't store exactly `math.MaxUint64` here, but only approximately).

jjo replied:
Same precision "loss" is present at the final
```
return float64(sample.Metric.Hash()) / float64MaxUint64
```
Given that strict number precision is not needed here, especially given the ~unpredictable  nature of `Metric.Hash() uint64` also (_unpredictable_ from the _PoV_ of the user who doesn't interact with its values), IMO it would not be an issue unless `ratio(1.0, metric)` wouldn't return all samples (verified with some `testdata/`).

juliusv replied:
Yes, that's what I also roughly arrived at - thanks for confirming :)

jjo replied:
cool yw+

juliusv reviewed (commented):
Looks pretty good overall, just a few comments

jjo committed:
Update promql/engine.go

Co-authored-by: Julius Volz <julius.volz@gmail.com>
Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
Update promql/engine.go

Co-authored-by: Julius Volz <julius.volz@gmail.com>
Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
Update promql/engine.go

Co-authored-by: Julius Volz <julius.volz@gmail.com>
Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented on docs/querying/operators.md:
```diff
@@ -230,6 +230,8 @@ vector of fewer elements with aggregated values:
 * `topk` (largest k elements by sample value)
 * `quantile` (calculate φ-quantile (0 ≤ φ ≤ 1) over dimensions)
+* `limitk` (sample n elements)
+* `limit_ratio` (sample elements with approximately 𝑟 ratio)
```
Good point, fwiw I see some interesting use-cases for `𝑟 < 0` (guaranteed to be the deterministic complement of `1 - (-r)`), like e.g. rejecting `avg` as a useful aggregation for your series, something like verifying that these averages are within reasonable values (_poor man's normality check_ if you wish): 
```
(avg(ratio(0.5, metric)) - avg(ratio(-0.5, metric))) < stddev(metric)
```
Adding some docu about it.

jjo replied:
Documented at ca498c1c, btw I'm not super-fused to this, happy to keep `𝑟  ≥ 0` only.

juliusv replied:
That's potentially interesting, yeah. Fine to keep it for now, as it's experimental anyway :)

jjo committed:
fix AddRatioSample() implementation to use a single conditional (instead of switch/case + fallback return)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
docs/querying/operators.md: document r < 0

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
add negative limit_ratio() example to docs/querying/examples.md

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented on docs/querying/examples.md:
```diff
@@ -95,3 +95,22 @@ Assuming this metric contains one time series per running instance, you could
+
+    limitk(10, app_foo_metric_bar)
+
+Alternatively, if we wanted the returned timeseries to be more evenly sampled,
```
I really like these new additional explanations and examples, but they are a bit out of place in this introductory PromQL examples page that otherwise only mentions a handful of the most popular language features. Explaining these details under https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators would be better IMO.

jjo replied:
Ah okies, will move.

jjo replied:
Moved+improved docu at 80e4463.

jjo replied:
Also added that _poor-man's-normality-check_ to `limit.test` at fe21d5a696a3dc0b369080bdb106ca18d1ffd36e.

juliusv commented on docs/querying/examples.md:
```diff
@@ -95,3 +95,22 @@ Assuming this metric contains one time series per running instance, you could
+
+    limit_ratio(-0.9, app_foo_metric_bar)
+
+
```
Too many empty lines at the end :)

jjo replied:
🤦🏼 moving them anyway off this file.

jjo committed:
move more extensive docu examples to docs/querying/operators.md

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
typo

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
small docu fix for poor-mans-normality-check, add it to limit.test ;)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
limit.test: expand "Poor man's normality check" to whole eval range

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
restore mistakenly removed existing small comment

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
expand poors-man-normality-check case(s)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented on docs/querying/operators.md:
```diff
@@ -291,6 +298,33 @@ To get the 5 largest HTTP requests counts across all instances we could write:
+You can also use this feature to e.g. verify that `avg()` is a representative
+aggregation for your samples' values, by checking that the difference between
+averaging two samples' subsets is "small" when compared to the standard
+deviation (and likely a great motivation to use `quantile()` instead ;)):
```
While I like humor, let's maybe keep the smileys out of the reference docs and keep the tone purely technical 😅

jjo replied:
Heh sure, undoing.

jjo replied:
undid at 3fab58e69 _remove humor from docs/querying/operators.md_.

jjo committed:
Revert "expand poors-man-normality-check case(s)"

This reverts commit f69e1603b2ebe69c0a100197cfbcf6f81644b564, indeed too
flaky 0:)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
remove humor from docs/querying/operators.md

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
fix signoff

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented:
Thanks, looks pretty good to me now. Before merging: I just noticed that the UI still needs updating (autocompletion and linting). Were you planning on doing that in this PR or a follow-up one?

Relevant places that will need to be updated for that (I hope I caught them all):

* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/parser/parser.ts#L170
* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L48-L61
* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L323-L344
* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/complete/promql.terms.ts#L583-L588
* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/highlight.js#L25
* https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/tokens.js#L60-L81

jjo committed:
add web/ui missing changes

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> Thanks, looks pretty good to me now. Before merging: I just noticed that the UI still needs updating (autocompletion and linting). Were you planning on doing that in this PR or a follow-up one?
> 
> Relevant places that will need to be updated for that (I hope I caught them all):
> 
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/parser/parser.ts#L170
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L48-L61
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L323-L344
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/complete/promql.terms.ts#L583-L588
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/highlight.js#L25
> * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/tokens.js#L60-L81

Oh cool, thanks for the pointers!, hopef00ly caught'em all at c4e9aa798 (mind some were already there).

jjo committed:
expand limit_ratio test cases, cross-fingering they'll not be flaky

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
remove flaky test

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented on promql/engine.go:
```diff
@@ -1345,6 +1346,23 @@ func (ev *evaluator) rangeEvalAgg(aggExpr *parser.AggregateExpr, sortedGrouping
+			return nil, warnings
+		case param < -1.0:
+			ratio = -1.0
+			warnings.Add(annotations.NewInvalidRatioWarning(param, ratio, aggExpr.Param.PositionRange()))
```
Hmm, I just tested this case, and there's no warning being returned from the API 🤔 I don't know why yet though, the `NewInvalidQuantileWarning` warning just below for `quantile()` is working just fine.

juliusv replied:
I've verified with Printfs that it's correctly entering this branch though, so now I wonder even more why the warning doesn't work...

juliusv replied:
Ah, it's because further below the case that handles the limiting operators does a return with a different subset of warnings without first merging in the other set:

https://github.com/prometheus/prometheus/blob/3d54bcc018df3e081758db3e13a49f9a2566c49b/promql/engine.go#L1365-L1370

I guess that return should be changed to:

```go
if ev.endTimestamp == ev.startTimestamp {
	warnings.Merge(ws)
	return result, warnings
}
```

jjo replied:
oh cool, thanks for thoroughly looking into it!, indeed a couple lines below there's such `Merge(ws)` following the flow -- added it at 883d080c3.

jjo committed:
add missing warnings.Merge(ws) in instant-query return shortcut

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
add missing LimitK||LimitRatio case to codemirror-promql/src/parser/parser.ts

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
> > Thanks, looks pretty good to me now. Before merging: I just noticed that the UI still needs updating (autocompletion and linting). Were you planning on doing that in this PR or a follow-up one?
> > Relevant places that will need to be updated for that (I hope I caught them all):
> > 
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/parser/parser.ts#L170
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L48-L61
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/promql.grammar#L323-L344
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/codemirror-promql/src/complete/promql.terms.ts#L583-L588
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/highlight.js#L25
> > * https://github.com/prometheus/prometheus/blob/a39e7e287f2173b456bbc8587bc103165c3d589e/web/ui/module/lezer-promql/src/tokens.js#L60-L81
> 
> Oh cool, thanks for the pointers!, hopef00ly caught'em all at ~[c4e9aa7](https://github.com/prometheus/prometheus/commit/c4e9aa79892e48e7a70e491b572d328fc08ded30)~ [272f9f7](https://github.com/prometheus/prometheus/pull/12503/commits/272f9f7b7cd36c10e72a2c609f23b423e79967b1) (mind some were already there).

BTW grepped `Topk` for places where `LimitK` (and `LimitRatio`) should be there also, fixed a missing one at 30215d16b

jjo committed:
fix ui-lint

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
7> > > Thanks, looks pretty good to me now. Before merging: I just noticed that the UI still needs updating (autocompletion and linting). Were you planning on doing that in this PR or a follow-up one?
> > > Relevant places that will need to be updated for that (I hope I caught them all):
>> [...]
> BTW grepped `Topk` for places where `LimitK` (and `LimitRatio`) should be there also, fixed a missing one at [30215d1](https://github.com/prometheus/prometheus/commit/30215d16b340005da093dfa38be998adf588a875)

Rendered output from
```
$ egrep --color=always -r 'Topk|LimitK' web/ui |egrep -v 'cjs|dist/|cache/'
```

<svg width="100" height="100" xmlns="http://www.w3.org/2000/svg">
<foreignObject width="100" height="100">
    <div xmlns="http://www.w3.org/1999/xhtml">
<pre><b class=MAG>web/ui/module/lezer-promql/src/tokens.js</b><b class=CYN>:</b>    <b class=HIR>LimitK</b>,
<b class=MAG>web/ui/module/lezer-promql/src/tokens.js</b><b class=CYN>:</b>    <b class=HIR>Topk</b>,
<b class=MAG>web/ui/module/lezer-promql/src/tokens.js</b><b class=CYN>:</b>    limitk: <b class=HIR>LimitK</b>,
<b class=MAG>web/ui/module/lezer-promql/src/tokens.js</b><b class=CYN>:</b>    topk: <b class=HIR>Topk</b>,
<b class=MAG>web/ui/module/lezer-promql/src/parser.js</b><b class=CYN>:</b>  nodeNames: "⚠ Bool Ignoring On GroupLeft GroupRight Offset Atan2 Avg Bottomk Count CountValues Group Max Min Quantile <b class=HIR>LimitK</b> LimitRatio Stddev Stdvar Sum <b class=HIR>Topk</b> By Without And Or Unless Start End LineComment PromQL AggregateExpr AggregateOp AggregateModifier GroupingLabels LabelName QuotedLabelName StringLiteral FunctionCallBody BinaryExpr Pow BoolModifier MatchingModifierClause Mul Div Mod Add Sub Eql Gte Gtr Lte Lss Neq FunctionCall FunctionIdentifier AbsentOverTime Identifier Absent Abs Acos Acosh Asin Asinh Atan Atanh AvgOverTime Ceil Changes Clamp ClampMax ClampMin Cos Cosh CountOverTime DaysInMonth DayOfMonth DayOfWeek DayOfYear Deg Delta Deriv Exp Floor HistogramCount HistogramFraction HistogramQuantile HistogramStdDev HistogramStdVar HistogramSum HistogramAvg HoltWinters Hour Idelta Increase Irate LabelReplace LabelJoin LastOverTime Ln Log10 Log2 MadOverTime MaxOverTime MinOverTime Minute Month Pi PredictLinear PresentOverTime QuantileOverTime Rad Rate Resets Round Scalar Sgn Sin Sinh Sort SortDesc SortByLabel SortByLabelDesc Sqrt StddevOverTime StdvarOverTime SumOverTime Tan Tanh Timestamp Time Vector Year MatrixSelector Duration NumberLiteral OffsetExpr ParenExpr SubqueryExpr UnaryExpr UnaryOp VectorSelector LabelMatchers UnquotedLabelMatcher MatchOp EqlSingle EqlRegex NeqRegex QuotedLabelMatcher StepInvariantExpr At AtModifierPreprocessors MetricName",
<b class=MAG>web/ui/module/lezer-promql/src/parser.js</b><b class=CYN>:</b>  <b class=HIR>LimitK</b> = 16,
<b class=MAG>web/ui/module/lezer-promql/src/parser.js</b><b class=CYN>:</b>  <b class=HIR>Topk</b> = 21,
<b class=MAG>web/ui/module/lezer-promql/src/promql.grammar</b><b class=CYN>:</b>  <b class=HIR>LimitK</b> |
<b class=MAG>web/ui/module/lezer-promql/src/promql.grammar</b><b class=CYN>:</b>  <b class=HIR>Topk</b>
<b class=MAG>web/ui/module/lezer-promql/src/promql.grammar</b><b class=CYN>:</b>  <b class=HIR>LimitK</b>,
<b class=MAG>web/ui/module/lezer-promql/src/promql.grammar</b><b class=CYN>:</b>  <b class=HIR>Topk</b>,
<b class=MAG>web/ui/module/lezer-promql/src/highlight.js</b><b class=CYN>:</b>    'Avg Bottomk Count Count_values Group <b class=HIR>LimitK</b> LimitRatio Max Min Quantile Stddev Stdvar Sum <b class=HIR>Topk</b>': tags.operatorKeyword,
<b class=MAG>web/ui/module/lezer-promql/src/parser.terms.js</b><b class=CYN>:</b>  <b class=HIR>LimitK</b> = 16,
<b class=MAG>web/ui/module/lezer-promql/src/parser.terms.js</b><b class=CYN>:</b>  <b class=HIR>Topk</b> = 21,
<b class=MAG>web/ui/module/lezer-promql/test/expression.txt</b><b class=CYN>:</b>          AggregateOp(<b class=HIR>Topk</b>),
<b class=MAG>web/ui/module/codemirror-promql/src/parser/parser.ts</b><b class=CYN>:</b>  <b class=HIR>LimitK</b>,
<b class=MAG>web/ui/module/codemirror-promql/src/parser/parser.ts</b><b class=CYN>:</b>  <b class=HIR>Topk</b>,
<b class=MAG>web/ui/module/codemirror-promql/src/parser/parser.ts</b><b class=CYN>:</b>      aggregateOp.type.id === <b class=HIR>Topk</b> ||
<b class=MAG>web/ui/module/codemirror-promql/src/parser/parser.ts</b><b class=CYN>:</b>      aggregateOp.type.id === <b class=HIR>LimitK</b> ||
</pre>
</foreignObject>
</svg>

juliusv commented on promql/engine.go:
```diff
@@ -1362,10 +1380,11 @@ func (ev *evaluator) rangeEvalAgg(aggExpr *parser.AggregateExpr, sortedGrouping
 			// If this could be an instant query, shortcut so as not to change sort order.
 			if ev.endTimestamp == ev.startTimestamp {
+				warnings.Merge(ws)
 				return result, ws
```
This will still have to be changed to return `warnings`, not `ws`.

jjo replied:
🤦🏼 ofc, fixed at 20235ee6c

jjo committed:
actually fix returned warnings :]

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

juliusv commented:
@jjo Thanks! :+1: Do things look good to merge from your side now?

jjo commented:
> @jjo Thanks! 👍 Do things look good to merge from your side now?

I think so (FWIW re-verified in a running instance that latest warnings fix is returned for both instant and range queries).

Thanks so much for the help with this! 🎉 :shipit:

juliusv commented:
Great! Then let's merge it, and we can always still adjust things since it's experimental and behind a feature flag :)

🚢
